w09h03/test/pgdp/filter/DGTest.java

```diff
@@ -0,0 +1,5539 @@
+        MyBufferedImageHeightNegative image = new MyBufferedImageHeightNegative(3, 3, BufferedImage.TYPE_INT_ARGB);
+        for(int i = 0; i < 3; i++){
+            for(int j = 0; j < 3; j++){
+                image.setRGB(i,j,Color.RED.getRGB());
+            }
+        }
+        MyFrameProvider provider = new MyFrameProvider(image);
+        //System.out.println(provider.getWidth());
+        //System.out.println(provider.getHeight());
+        try{
+            VideoContainer container = new VideoContainer(provider);
+        }
+        catch(IllegalVideoFormatException ivfe){
+            System.out.println("correct exception thrown!");
+        }
+        catch(Exception e){
+            Assertions.fail();
+        }
+    }
+
+    @Test
+    @DisplayName("Test DG 106 VideoContainer Konstruktor both 0")
+    public void test106() {
+        MyBufferedImageBothZero image = new MyBufferedImageBothZero(3, 3, BufferedImage.TYPE_INT_ARGB);
+        for(int i = 0; i < 3; i++){
+            for(int j = 0; j < 3; j++){
+                image.setRGB(i,j,Color.RED.getRGB());
+            }
+        }
+        MyFrameProvider provider = new MyFrameProvider(image);
+        //System.out.println(provider.getWidth());
+        //System.out.println(provider.getHeight());
+        try{
+            VideoContainer container = new VideoContainer(provider);
+        }
+        catch(IllegalVideoFormatException ivfe){
+            System.out.println("correct exception thrown!");
+        }
+        catch(Exception e){
+            Assertions.fail();
+        }
+    }
+
+    @Test
+    @DisplayName("Test DG 107 VideoContainer Konstruktor both negative")
+    public void test107() {
+        MyBufferedImageBothNegative image = new MyBufferedImageBothNegative(3, 3, BufferedImage.TYPE_INT_ARGB);
+        for(int i = 0; i < 3; i++){
+            for(int j = 0; j < 3; j++){
+                image.setRGB(i,j,Color.RED.getRGB());
+            }
+        }
+        MyFrameProvider provider = new MyFrameProvider(image);
+        //System.out.println(provider.getWidth());
+        //System.out.println(provider.getHeight());
+        try{
+            VideoContainer container = new VideoContainer(provider);
+        }
+        catch(IllegalVideoFormatException ivfe){
+            System.out.println("correct exception thrown!");
+        }
+        catch(Exception e){
+            Assertions.fail();
+        }
+    }
+
+    // VIDEOCONTAINER LIMIT TESTS
+
+    @Test
+    @DisplayName("Test DG 120 VideoContainer Limit noot auf 100 Frames")
+    public void test120() throws FileNotFoundException, IllegalVideoFormatException {
+
+        int length = 100;
+
+        FrameProvider fp = new FrameProvider("noot.mp4");
+        FrameProvider fp1 = new FrameProvider("noot.mp4");
+        VideoContainer in = new VideoContainer(fp);
+        VideoContainer in1 = new VideoContainer(fp1);
+
+        Object[] array = in1.frameStream.limit(length).toArray();
```

Comment:
`frameStream` is private in the artemis template cannot be accessed directly.

Reply:
Code from the Artemis Template
```java
public class VideoContainer {

	private FrameProvider provider;

	private Stream<Frame> frameStream;
	
	/**
	 * Nutzt javacv um Videodatei darzustellen.
	 */
	public VideoContainer(FrameProvider fp) {

		// TODO: Implementieren
		throw new NotImplementedException();
	}
	```

Reply:
Yes, possible, you can create a PR here and suggest the changes. As written before, the code is not ideal, but consists of extensives tests. The advantages by that out-weight the imperfections. 
I'd say: Either you create the changes yourself or we merge this one, since the functionality of the tests is what matters to 99% of the students, not the beauty of the code.